imports/lib/models/Model.ts

```diff
@@ -0,0 +1,187 @@
+    newShape[key] = fieldSchema.isOptional() ? fieldSchema : fieldSchema.optional();
+  }
+
+  return new z.ZodObject({
+    ...schema._def,
+    shape: () => newShape,
+  });
+}
+
+const modifierIsWholeDoc = <T extends Document>(modifier: Mongo.Modifier<T>): modifier is T => {
+  return !Object.keys(modifier).some((k) => k.startsWith('$'));
+};
+
+export const AllModels = new Set<Model<any>>();
+
+class Model<T extends z.ZodObject<any, 'strict'>> {
+  name: string;
+
+  schema: T;
+
+  collection: Mongo.Collection<z.output<T>>;
+
+  constructor(name: string, schema: T) {
+    this.name = name;
+    this.schema = schema;
+    this.collection = new Mongo.Collection(name);
+    AllModels.add(this);
+  }
+
+  async insertAsync(doc: z.input<T>): Promise<z.output<T>['_id']> {
+    const cleaned: z.output<T> = await IsInsert.withValue(true, async () => {
+      return this.schema.parseAsync(doc);
+    });
+    return this.collection.insert(cleaned as any) as any;
+  }
+
+  // As a note, we don't accept full documents as modifiers because our
+  // transform functions don't make sense (e.g. we won't accept a `createdAt`
+  // timestamp, but we expect one to be there)
+  async updateAsync(
+    selector: Selector<z.output<T>>,
+    modifier: Exclude<Mongo.Modifier<z.input<T>>, z.input<T>>,
+    options: {
+      multi?: boolean | undefined;
+      upsert?: boolean | undefined;
+      arrayFilters?: { [identifier: string]: any }[] | undefined;
+    } = {},
+  ): Promise<number> {
+    // Types should prevent passing a full document as a modifier, but verify at
+    // runtime to be sure
+    const rawModifier: Mongo.Modifier<z.input<T>> = modifier as any;
+    if (modifierIsWholeDoc(rawModifier)) {
+      throw new Error('Cannot update with a full document');
+    }
+
+    const setFields = await IsUpdate.withValue(!options.upsert, async () => {
+      return optionalPreservingPartial(this.schema).parseAsync(rawModifier.$set ?? {});
+    });
+    const setOnInsertFields = await IsUpsert.withValue(true, async () => {
+      const cleaned = await optionalPreservingPartial(this.schema)
+        .parseAsync(rawModifier.$setOnInsert ?? {});
+      return Object.fromEntries(Object.entries(cleaned).filter(([k]) => !(k in setFields)));
+    });
+
+    const mergedModifier: Mongo.Modifier<z.output<T>> = {
+      ...rawModifier,
+      $setOnInsert: {
+        ...setOnInsertFields,
+      },
+      $set: {
+        ...setFields,
+      },
+    };
+
+    return this.collection.update(selector, mergedModifier, options);
+  }
+
+  async upsertAsync(
+    selector: Selector<z.output<T>>,
+    modifier: Exclude<Mongo.Modifier<z.output<T>>, z.output<T>>,
+    options: {
+      multi?: boolean | undefined;
+    } = {},
+  ): Promise<{
+    numberAffected?: number | undefined;
+    insertedId?: z.output<T>['_id'] | undefined;
+  }> {
+    const rawModifier: Mongo.Modifier<T> = modifier as any;
+    if (modifierIsWholeDoc(rawModifier)) {
```

Comment:
Note: I think we do have one `upsertAsync` call that uses an empty doc as the modifier: https://github.com/deathandmayhem/jolly-roger/blob/5e3161cb99da5ae48dcf9ee1a2a852eeaa931a2d/imports/server/addUsersToDiscordRole.ts#L65-L70

Reply:
Ah, OK, so `modifierIsWholeDoc` should just return true if the object is empty.